src/app/content/highlights/components/Card.tsx

```diff
@@ -54,94 +54,117 @@ export interface CardProps {
-const Card = (props: CardProps) => {
+type CardPropsWithBookAndPage = Omit<CardProps, 'book' | 'page'> & {
+  book: Exclude<CardProps['book'], undefined>;
+  page: Exclude<CardProps['page'], undefined>;
+};
+
+type ComputedProps = {
+  focusCard(): Promise<void>;
+  editing: boolean;
+  setEditing: React.Dispatch<React.SetStateAction<boolean>>;
+  annotation: string | undefined;
+  hasUnsavedHighlight: boolean;
+  element: React.RefObject<HTMLElement>
+};
+
+function useComputedProps(props: CardProps): ComputedProps {
   const annotation = props.data && props.data.annotation;
-  const element = React.useRef<HTMLElement>(null);
+  const services = useServices();
+  const { isActive, highlight, focus, onHeightChange,
+lastActive } = props;
+  const {id} = highlight;
   const [editing, setEditing] = React.useState<boolean>(!annotation);
-  const [highlightRemoved, setHighlightRemoved] = React.useState<boolean>(false);
-  const locationFilters = useSelector(selectHighlights.highlightLocationFilters);
   const hasUnsavedHighlight = useSelector(selectHighlights.hasUnsavedHighlight);
-  const services = useServices();
-
-  const { isActive, highlight: { id }, focus } = props;
-
   const focusCard = React.useCallback(async() => {
     if (!isActive && (!hasUnsavedHighlight || await showConfirmation(services))) {
       focus(id);
     }
   }, [isActive, hasUnsavedHighlight, id, focus, services]);
+  const element = React.useRef<HTMLElement>(null);
 
   useFocusIn(element, true, focusCard);
 
   React.useEffect(() => {
-    if (!props.lastActive) {
+    if (!lastActive) {
       setEditing(false);
     } else {
-      scrollHighlightIntoView(props.highlight, element);
+      scrollHighlightIntoView(highlight, element);
     }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [element, props.lastActive]);
+  }, [element, lastActive, highlight]);
 
   React.useEffect(() => {
     if (annotation) {
-      props.highlight.elements.forEach((el) => (el as HTMLElement).classList.add('has-note'));
+      highlight.elements.forEach((el) => (el as HTMLElement).classList.add('has-note'));
     } else {
-      props.highlight.elements.forEach((el) => (el as HTMLElement).classList.remove('has-note'));
+      highlight.elements.forEach((el) => (el as HTMLElement).classList.remove('has-note'));
     }
-  }, [props.highlight, annotation]);
+  }, [highlight, annotation]);
 
   React.useEffect(() => {
-    if (!annotation && !props.isActive) {
-      props.onHeightChange({ current: null });
+    if (!annotation && !isActive) {
+      onHeightChange({ current: null });
     }
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [annotation, props.isActive]);
+  }, [annotation, isActive]); // onHeightChange
+
+return React.useMemo(
+    () => ({
+      focusCard,
+      editing,
+      setEditing,
+      annotation,
+      hasUnsavedHighlight,
+      element,
+    }),
+    [annotation, editing, focusCard, hasUnsavedHighlight]
```

Comment:
Don't you need setEditing and element in the array?

Reply:
State set functions are known not to change, and refs are explicitly not change-tracked.

Reply:
roy is right, but you may see us listing dependencies in other places because the linter has special cases to not complain if you're using the result of `useRef` or `useState`, but if you're composing those things through a helper hook the linter can't keep up. its the same thing either way because they never update, but you end up needing to specify a bunch of extra dependencies sometimes.